File: rust/src/bin/p0049.rs
// Project Euler: Problem 49

use std::collections::HashMap;

euler::run_solver!(49);

fn solve() -> String {
    compute(4).to_string()
}

fn compute(n_digits: u32) -> i64 {
    let p_tbl = get_prime_tbl(n_digits);
    for lst in p_tbl.values() {
        let length = lst.len();
        if length < 3 {
            continue;
        }

        for i in 0..(length - 2) {
            for j in (i + 1)..(length - 1) {
                let tmp = lst[j] * 2 - lst[i];
                if lst.contains(&tmp) && lst[i] != 1487 {
                    return [lst[i], lst[j], tmp]
                        .into_iter()
                        .fold(0, |acc, x| acc * 10_i64.pow(n_digits) + x);
                }
            }
        }
    }

    // Not reached on this problem
    unreachable!();
}

fn get_prime_tbl(n_digits: u32) -> HashMap<i64, Vec<i64>> {
    use euler::math::{self, primes};
    use std::cmp;

    let mut p_tbl: HashMap<i64, Vec<i64>> = HashMap::new();

    for p in primes::primes(10_i64.pow(n_digits - 1), 10_i64.pow(n_digits)) {
        let mut key_digits = math::digits(p);
        key_digits.sort_by_key(|&x| cmp::Reverse(x));
        let key = math::undigits(&key_digits);
        let mut lst = match p_tbl.get(&key) {
            Some(v) => v.clone(),
            None => Vec::new(),
        };
        lst.push(p);
        p_tbl.insert(key, lst);
    }
    p_tbl
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0049_ndigits_4() {
        assert_eq!(compute(4), 296962999629);
    }
}

File: ocaml/src/p0049.ml
(* Project Euler: Problem 49 *)

open Core

let no_exist_msg = "There are no such sequences."

let get_prime_group_lst n_digits =
  let make_key n =
    Euler.Util.(digits n |> List.sort ~compare:Int.ascending |> undigits)
  in

  Euler.Math.Prime.primes (Int.pow 10 (n_digits - 1)) (Int.pow 10 n_digits)
  |> List.map ~f:(fun n -> (make_key n, n))
  |> Euler.Util.list_assoc_group
  |> List.map ~f:snd
;;

let compute n_digits =
  assert (n_digits > 0 && n_digits <= 4);

  let result =
    List.(
      get_prime_group_lst n_digits
      |> filter_map ~f:(fun lst ->
        if length lst < 3 then None else Some (sort ~compare:Int.ascending lst))
      >>= Euler.Util.combination 3
      |> filter ~f:(fun triplet ->
        nth_exn triplet 1 - nth_exn triplet 0 = nth_exn triplet 2 - nth_exn triplet 1)
      |> map ~f:(Fun.compose (reduce_exn ~f:( ^ )) (map ~f:Int.to_string))
      |> filter ~f:(Fun.compose not (String.equal "148748178147")))
  in
  match List.length result with
  | 0 -> no_exist_msg
  | 1 -> List.hd_exn result
  | _ -> failwith "Error: There are two or more sequences."
;;

let solve () = compute 4

(* Test *)

let%test_unit "1" = [%test_eq: string] (compute 1) no_exist_msg
let%test_unit "2" = [%test_eq: string] (compute 2) no_exist_msg
let%test_unit "3" = [%test_eq: string] (compute 3) no_exist_msg
let%test_unit "4" = [%test_eq: string] (compute 4) "296962999629"
